src/v/config/property.h

```diff
@@ -123,6 +132,152 @@ class property : public base_property {
+        return "seed_server";
+    } else if constexpr (std::is_same_v<type, unresolved_address>) {
+        return "unresolved_address";
+    } else if constexpr (std::is_same_v<type, tls_config>) {
+        return "tls_config";
+    } else if constexpr (std::is_same_v<type, endpoint_tls_config>) {
+        return "endpoint_tls_config";
+    } else if constexpr (std::is_same_v<type, model::broker_endpoint>) {
+        return "broker_endpoint";
+    } else if constexpr (std::is_floating_point_v<type>) {
+        return "number";
+    } else if constexpr (std::is_integral_v<type>) {
+        return "integer";
+    }
+
+    // Never used, the `property_type_name` wrapper requires that
+    // this value is not returned when instantiating the function for
+    // any type.
+    return "";
+}
+
+// Wrapper that uses a `requires` to ensure that method is only instantiated
+// if property_type_name_inner can name the type: will generate compile error
+// if someone creates a property<> instantiation that the type name code
+// doesn't understand.
+// This is better than a static_assert because the compile error will name
+// the instantiated type, rather than just nonspecifically noting an
+// assertion failure.
+template<typename T>
+requires(property_type_name_inner<T>() != "") consteval std::string_view
+  property_type_name() {
+    return property_type_name_inner<T>();
+}
+
+template<typename T>
+std::string_view property<T>::type_name() const {
+    // Go via a non-member function so that specialized implementations
+    // can use concepts without all the same concepts having to apply
+    // to the type T in the class definition.
+    return property_type_name<T>();
+}
+
+template<typename T>
+consteval std::optional<std::string_view> property_units_name() {
+    using type = std::decay_t<T>;
+    if constexpr (std::is_same_v<type, std::chrono::milliseconds>) {
+        return "ms";
+    } else if constexpr (std::is_same_v<type, std::chrono::seconds>) {
+        return "s";
+    } else if constexpr (reflection::is_std_optional_v<type>) {
+        return property_units_name<typename type::value_type>();
+    } else {
+        return std::nullopt;
+    }
+}
+
+inline std::optional<std::string_view>
+property_units_name(const std::chrono::milliseconds*) {
+    return "ms";
+}
+
+inline std::optional<std::string_view>
+property_units_name(const std::chrono::seconds*) {
+    return "s";
+}
+
+// Default units is null -- e.g. any non-numeric setting
+template<typename T>
+inline std::optional<std::string_view> property_units_name(T*) {
+    return std::nullopt;
+}
+
+template<typename T>
+requires reflection::is_std_optional_v<std::decay_t<T>>
+  std::optional<std::string_view> property_units_name(T*) {
+    typename T::value_type const* ptr = nullptr;
+    return property_units_name(ptr);
+}
+
+template<typename T>
+std::optional<std::string_view> property<T>::units_name() const {
+    return property_units_name(&_value);
+}
```

Comment:
I'm probably missing something, but:

```suggestion
template<typename T>
std::optional<std::string_view> property<T>::units_name() const {
    return property_units_name<T>();
}
```

Reply:
I left the property_units_name() in there by accident... it was crashing clang.

Replaced this with a compromise version that uses the `if constexpr` style but avoids returning a std::optional from the constexpr function, which appears to be what triggered the compiler barf.